packages/accounts-controller/src/AccountsController.ts

```diff
@@ -0,0 +1,580 @@
+
+export type AccountsControllerState = {
+  internalAccounts: {
+    accounts: Record<string, InternalAccount>;
+    selectedAccount: string; // id of the selected account
+  };
+};
+
+export type AccountsControllerGetStateAction = {
+  type: `${typeof controllerName}:getState`;
+  handler: () => AccountsControllerState;
+};
+
+export type AccountsControllerActions = AccountsControllerGetStateAction;
+
+export type AccountsControllerChangeEvent = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [AccountsControllerState, Patch[]];
+};
+
+export type AccountsControllerSelectedAccountChangeEvent = {
+  type: `${typeof controllerName}:selectedAccountChange`;
+  payload: [InternalAccount];
+};
+
+export type AccountsControllerEvents =
+  | AccountsControllerChangeEvent
+  | AccountsControllerSelectedAccountChangeEvent
+  | SnapControllerEvents
+  | KeyringControllerEvents;
+
+export type AccountsControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  AccountsControllerActions,
+  AccountsControllerEvents,
+  string,
+  string
+>;
+
+const accountsControllerMetadata = {
+  internalAccounts: {
+    persist: true,
+    anonymous: false,
+  },
+  selectedAccount: {
+    persist: true,
+    anonymous: false,
+  },
+};
+
+const defaultState: AccountsControllerState = {
+  internalAccounts: {
+    accounts: {},
+    selectedAccount: '',
+  },
+};
+
+/**
+ * Controller that manages internal accounts.
+ * The accounts controller is responsible for creating and managing internal accounts.
+ * It also provides convenience methods for accessing and updating the internal accounts.
+ * The accounts controller also listens for keyring state changes and updates the internal accounts accordingly.
+ * The accounts controller also listens for snap state changes and updates the internal accounts accordingly.
+ *
+ */
+export class AccountsController extends BaseControllerV2<
+  typeof controllerName,
+  AccountsControllerState,
+  AccountsControllerMessenger
+> {
+  getKeyringForAccount: KeyringController['getKeyringForAccount'];
+
+  getKeyringByType: KeyringController['getKeyringsByType'];
+
+  getAccounts: KeyringController['getAccounts'];
+
+  keyringApiEnabled: boolean;
+
+  /**
+   * Constructor for AccountsController.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The messenger object.
+   * @param options.state - Initial state to set on this controller
+   * @param [options.keyringApiEnabled] - The keyring API enabled flag.
+   * @param options.getKeyringForAccount - Gets the keyring for a given account from the keyring controller.
+   * @param options.getKeyringByType - Gets the keyring instance for the given type from the keyring controller.
+   * @param options.getAccounts - Gets all the accounts from the keyring controller.
+   */
+  constructor({
+    messenger,
+    state,
+    keyringApiEnabled,
+    getKeyringForAccount,
+    getKeyringByType,
+    getAccounts,
```

Comment:
We could use messenger actions to avoid passing controller instances or methods around in other controllers' constructor parameters.

These actions are not available at the moment in `KeyringController`'s messenger API, but if that makes sense to you I can help add them: This would allow us to remove these three constructor parameters.

Reply:
Makes sense, i've made a change on the keyring controller to add the needed actions.